windows-gaming-driver/src/setup/vfio.rs

```diff
@@ -91,84 +88,78 @@ fn select(setup: &mut SetupConfig, machine: &mut MachineConfig) -> Result<()> {
+		println!("If you kill your system using this there is only one thing i can say to you:");
+		println!();
+		println!("U done goofed m8.");
+		println!("Ur on ur own :)");
+		println!();
+		let all_devices = pci_devs.iter().filter(|x| x.pci_class != "60400").collect();
+		while let Ok(Some(())) = select_device(setup, machine, "Which device do you want to pass through?", &all_devices) {};
+	}
     Ok(())	
 }
 
-fn ask_selection<'a>(question: &str, filter: &Fn(&PciDevice) -> (bool), pci_devices: Vec<PciDevice<'a>>, permanent: bool) -> (Vec<VfioDevice>, Vec<PciId>, Vec<PciDevice<'a>>) {
-    
-    let (mut devices, remaining_pci_devices) = pci_devices.into_two(filter);
-    
-    println!();
-    
-    for (i, dev) in devices.iter().enumerate() {
-        println!("[{}]\t{}", i, dev);
-    }
-	println!("[{}]\tNone of the above.", devices.len());
-
-    let selection = ask::numeric(question, 0..devices.len()+1);
-    
-    if selection >= devices.len() {
-	    devices.extend(remaining_pci_devices);
-    	return (Vec::new(), Vec::new(), devices);
-    }
-    let selected = devices.remove(selection);
-
-	devices.extend(remaining_pci_devices);
+fn select_device(setup: &mut SetupConfig, machine: &mut MachineConfig, question: &str, devices: &Vec<&PciDevice>) -> Result<Option<()>>{
+    let udev = Context::new().expect("Failed to create udev context");
+    let mut iter = Enumerator::new(&udev)?;
+    iter.match_subsystem("pci")?;
+    let udev_devices = iter.scan_devices()?.map(PciDevice::new);
+    let pci_devs: Vec<_> = udev_devices.filter(|x| !machine.vfio_slots.iter().any(|y| y.device == x.id)).collect();
+    let askable_devices: Vec<_> = devices.iter().filter(|x| !machine.vfio_slots.iter().any(|y| y.device == x.id)).collect();
 	
-	let (mut related_devices, unpassed_pci_devices) = devices.into_two(&|dev| dev.pci_device() == selected.pci_device());
-	related_devices.push(selected);
+	println!("{}", devices.len());
 	
-	let (mut ret_devices, mut ret_devids, ret_pci_devices) = ask_selection(question, filter, unpassed_pci_devices, permanent);
-	ret_devices.extend(related_devices.iter().map(|dev| if permanent {VfioDevice::Permanent(dev.pci_slot.clone())} else {VfioDevice::Temporarily(dev.pci_slot.clone())}));
-	ret_devids.extend(related_devices.iter().map(|dev| dev.id));
+	for (i, dev) in askable_devices.iter().enumerate() {
+	    println!("[{}]\t{}", i, dev);
+	}
+	println!("[{}]\tNone of the above.", askable_devices.len());
+
+	let selection = ask::numeric(question, 0..askable_devices.len()+1);
 	
-	(ret_devices, ret_devids, ret_pci_devices)
+	if selection < askable_devices.len() {
+		let ref selected = askable_devices[selection];
+		for device in pci_devs.iter().filter(|x| x.pci_device() == selected.pci_device()) {
+			let vfio_device = VfioDevice {
+				resettable: device.resettable,
+				slot: device.pci_slot.clone(),
+				device: device.id,
+			};
+			
+			if !device.resettable {
+				setup.vfio_devs.push(device.id);
```

Comment:
In this code we can, however at least iommu.rs needs it later on.

You *could* argue that the vfio-module should just do the iommu checking directly instead of putting it into a different step, which could help, if you already have iommu enabled for some reason and we could've checked it when you first select the vfio-devices. But i think that is an entirely different issue.